synapse/storage/databases/main/schema/delta/60/01refresh_tokens.sql.postgres

```diff
@@ -0,0 +1,13 @@
+CREATE TABLE refresh_tokens (
+  id BIGINT PRIMARY KEY,
+  user_id TEXT NOT NULL,
+  device_id TEXT,
+  token TEXT NOT NULL,
+  next_token_id BIGINT REFERENCES refresh_tokens (id) ON DELETE CASCADE,
```

Comment:
this is the reason the port script is getting stuck: it tries to populate the tables in a way that constraints will be satisfied, by populating the least-constrained tables first. The logic is being confused by `refresh_tokens`'s reference to itself.

I've not looked at the rest of the code yet, but one solution might be simply to add `AND tc.table_name != ccu.table_name` to the `get_table_constraints` query in the port script.

Reply:
Will try that, thanks!

Reply:
Done in https://github.com/matrix-org/synapse/pull/9450/commits/29806b4f1a18ccb56848dd66183ca825a7f25964